File: android/src/main/java/com/example/flutter_lottie/LottieView.java
package com.example.flutter_lottie;

import android.animation.Animator;
import android.content.Context;
import android.graphics.Color;
import android.view.View;
import com.airbnb.lottie.LottieAnimationView;
import com.airbnb.lottie.LottieProperty;
import com.airbnb.lottie.model.KeyPath;
import com.airbnb.lottie.value.LottieValueCallback;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import io.flutter.plugin.platform.PlatformView;
import java.util.Map;

public class LottieView implements PlatformView, MethodChannel.MethodCallHandler {
    private final Context mContext;
    private final int mId;
    private final Object mArgs;
    private final Registrar mRegistrar;
    private final LottieAnimationView animationView;

    private float maxFrame;
    private MethodChannel channel;
    private EventChannel.EventSink onPlaybackFinishEvent;

    LottieView(Context context, int id, Object args, Registrar registrar) {
        super();
        mContext = context;
        mId = id;
        mArgs = args;
        mRegistrar = registrar;
        animationView = new LottieAnimationView(context);

        Map<String, Object> params = (Map<String, Object>) args;
        create(params);

    }

    void create(Map<String, Object> args) {

        channel = new MethodChannel(mRegistrar.messenger(), "convictiontech/flutter_lottie_" + mId);
        channel.setMethodCallHandler(this);

        final EventChannel onPlaybackCompleteEventChannel = new EventChannel(mRegistrar.messenger(), "convictiontech/flutter_lottie_stream_playfinish_" + mId);

        onPlaybackCompleteEventChannel.setStreamHandler(new EventChannel.StreamHandler() {
            @Override
            public void onListen(Object o, EventChannel.EventSink eventSink) {
                onPlaybackFinishEvent = eventSink;
            }

            @Override
            public void onCancel(Object o) {

            }
        });


        if(args.get("url") != null) {
            animationView.setAnimationFromUrl(args.get("url").toString());
        }

        if(args.get("filePath") != null) {
            String key = mRegistrar.lookupKeyForAsset(args.get("filePath").toString());
            animationView.setAnimation(key);
        }

        boolean loop, reverse, autoPlay;

        loop = ((args.get("loop")) != null) ? Boolean.parseBoolean(args.get("loop").toString()) : false;
        reverse = ((args.get("reverse")) != null) ? Boolean.parseBoolean(args.get("reverse").toString()) : false;
        autoPlay = ((args.get("autoPlay")) != null) ? Boolean.parseBoolean(args.get("autoPlay").toString()) : false;

        animationView.setRepeatCount(loop ? -1 : 0);

        maxFrame = animationView.getMaxFrame();

        if(reverse) {
            animationView.setRepeatMode(2);
        } else {
            animationView.setRepeatMode(1);
        }

        if(autoPlay) {
            animationView.playAnimation();
        }


        animationView.addAnimatorListener(new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animation) {
            }

            @Override
            public void onAnimationEnd(Animator animation) {
                if(onPlaybackFinishEvent != null) {
                    onPlaybackFinishEvent.success(true);
                }
            }

            @Override
            public void onAnimationCancel(Animator animation) {
                if(onPlaybackFinishEvent != null) {
                    onPlaybackFinishEvent.success(false);
                }
            }

            @Override
            public void onAnimationRepeat(Animator animation) {
            }
        });

    }

    @Override
    public View getView() {
        return animationView;
    }

    @Override
    public void dispose() {
        animationView.cancelAnimation();
    }

    @Override
    public void onMethodCall(MethodCall call, MethodChannel.Result result) {
        Map<String, Object> args = (Map<String, Object>) call.arguments;
        switch(call.method) {
            case "play":
                animationView.setMinAndMaxFrame(0, (int) maxFrame);
                animationView.setMinAndMaxProgress(0, 1);
                animationView.playAnimation();
                break;
            case "resume":
                animationView.resumeAnimation();
                break;
            case "playWithProgress":
                if(args.containsKey("fromProgress") && args.get("fromProgress") != null) {
                    final float fromProgress = ((Double) args.get("fromProgress")).floatValue();
                    animationView.setMinProgress(fromProgress);
                }

                if(args.get("toProgress") != null) {
                    final float toProgress = ((Double) args.get("toProgress")).floatValue();
                    animationView.setMaxProgress(toProgress);
                }
                animationView.playAnimation();
                break;
            case "playWithFrames":
                if(args.get("fromFrame") != null) {
                    final int fromFrame = (int) args.get("fromFrame");
                    animationView.setMinFrame(fromFrame);
                }

                if(args.get("toFrame") != null) {
                    final int toFrame = (int) args.get("toFrame");
                    animationView.setMaxFrame(toFrame);
                }
                animationView.playAnimation();
                break;
            case "stop":
                animationView.cancelAnimation();
                animationView.setProgress(0.0f);
                final int mode = animationView.getRepeatMode();
                animationView.setRepeatMode(1);
                animationView.setRepeatMode(mode);
                break;
            case "pause":
                animationView.pauseAnimation();
                break;
            case "setAnimationSpeed":
                animationView.setSpeed(Float.parseFloat(args.get("speed").toString()));
                break;
            case "setLoopAnimation":
                boolean loop = ((args.get("loop")) != null) ? Boolean.parseBoolean(args.get("loop").toString()) : false;
                animationView.setRepeatCount(loop ? -1 : 0);
                break;
            case "setAutoReverseAnimation":
                boolean reverse = ((args.get("reverse")) != null) ? Boolean.parseBoolean(args.get("reverse").toString()) : false;
                if(reverse) {
                    animationView.setRepeatMode(2);
                } else {
                    animationView.setRepeatMode(1);
                }
                break;
            case "setAnimationProgress":
                animationView.pauseAnimation(); // TODO Make sure its consistant with iOS
                animationView.setProgress(Float.parseFloat(args.get("progress").toString()));
                break;
            case "setProgressWithFrame":
                animationView.setFrame(Integer.parseInt(args.get("progress").toString()));
                break;
            case "isAnimationPlaying":
                result.success(animationView.isAnimating());
                break;
            case "getAnimationDuration":
                result.success((double)animationView.getDuration());
                break;
            case "getAnimationProgress":
                result.success((double)animationView.getProgress());
                break;
            case "getAnimationSpeed":
                result.success((double)animationView.getSpeed());
                break;
            case "getLoopAnimation":
                result.success(animationView.getRepeatCount() == -1 ? true : false);
                break;
            case "getAutoReverseAnimation":
                result.success(animationView.getRepeatMode() == 2 ? true : false);
                break;
            case "setValue":
                final String value = args.get("value").toString();
                final String keyPath = args.get("keyPath").toString();
                final String type = args.get("type").toString();
                setValue(type, value, keyPath);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    private void setValue(String type, String value, String keyPath) {
        String[] keyPaths = keyPath.split("\\.");
        switch (type) {
            case "LOTColorValue":
                animationView.addValueCallback(
                        new KeyPath(keyPaths),
                        LottieProperty.COLOR,
                        new LottieValueCallback<>(convertColor(value))
                );
                break;
            case "LOTOpacityValue":
                float v = Float.parseFloat(value) * 100;
                animationView.addValueCallback(
                    new KeyPath(keyPaths),
                    LottieProperty.OPACITY,
                    new LottieValueCallback<>(Math.round(v))
                );
                break;
            default:
                break;
        }
    }

//    private int getOpacity(String value) {
//        return Integer.valueOf(value.substring(2,4), 16);
//    }

    private int convertColor(String value) {
//        float alpha = Integer.valueOf(value.substring(2,4), 16);
        float red = Integer.valueOf(value.substring(4,6), 16);
        float green = Integer.valueOf(value.substring(6,8), 16);
        float blue = Integer.valueOf(value.substring(8,10), 16);
        return Color.argb(255, (int) red, (int) green, (int) blue);
    }
}

File: ios/Classes/LottieView.swift
import UIKit
import Flutter
import Lottie

public class LottieView : NSObject, FlutterPlatformView {
   let frame : CGRect
   let viewId : Int64
   
   var animationView: LOTAnimationView?
   var testStream : TestStreamHandler?
   var delegates : [LOTValueDelegate]
   var registrarInstance : FlutterPluginRegistrar
   
   
   init(_ frame: CGRect, viewId: Int64, args: Any?, registrarInstance : FlutterPluginRegistrar) {
      self.frame = frame
      self.viewId = viewId
      self.registrarInstance = registrarInstance
      self.delegates = []
      
      super.init()
      
      self.create(args: args)
   }
   
   func create(args: Any?) {
      
      let channel : FlutterMethodChannel = FlutterMethodChannel.init(name: "convictiontech/flutter_lottie_" + String(viewId), binaryMessenger: self.registrarInstance.messenger())
      let handler : FlutterMethodCallHandler = methodCall;
      channel.setMethodCallHandler(handler)
      
      let testChannel = FlutterEventChannel(name: "convictiontech/flutter_lottie_stream_playfinish_"  + String(viewId), binaryMessenger: self.registrarInstance.messenger())
      self.testStream  = TestStreamHandler()
      testChannel.setStreamHandler(testStream as? FlutterStreamHandler & NSObjectProtocol)
      
      
      if let argsDict = args as? Dictionary<String, Any> {
         let url = argsDict["url"] as? String ?? nil;
         let filePath = argsDict["filePath"] as? String ?? nil;
         
         if url != nil {
            self.animationView = LOTAnimationView(contentsOf: URL(string: url!)!)
         }
         
         if filePath != nil {
            print("THIS IS THE ID " + String(viewId) + " " + filePath!)
            let key = self.registrarInstance.lookupKey(forAsset: filePath!)
            let path = Bundle.main.path(forResource: key, ofType: nil)
            self.animationView = LOTAnimationView(filePath: path!)
         }
         
         let loop = argsDict["loop"] as? Bool ?? false
         let reverse = argsDict["reverse"] as? Bool ?? false
         let autoPlay = argsDict["autoPlay"] as? Bool ?? false
         
         
         self.animationView?.loopAnimation = loop
         self.animationView?.autoReverseAnimation = reverse
         self.animationView?.completionBlock = completionBlock;
         if(autoPlay) {
            self.animationView?.play(completion: completionBlock);
         }
         
      }
      
   }
   
   public func view() -> UIView {
      return animationView!
   }
   
   public func completionBlock(animationFinished : Bool) -> Void {
      if let ev : FlutterEventSink = self.testStream!.event {
         ev(animationFinished)
      }
   }
   
   
   func methodCall( call : FlutterMethodCall, result: FlutterResult ) {
      var props : Dictionary<String, Any>  = [String: Any]()
      
      if let args = call.arguments as? Dictionary<String, Any> {
         props = args
      }
      
      if(call.method == "play") {
         self.animationView?.animationProgress = 0
         self.animationView?.play(completion: completionBlock);
      }
      
      if(call.method == "resume") {
         self.animationView?.play(completion: completionBlock);
      }
      
      if(call.method == "playWithProgress") {
         let toProgress = props["toProgress"] as! CGFloat;
         if let fromProgress = props["fromProgress"] as? CGFloat {
            self.animationView?.play(fromProgress: fromProgress, toProgress: toProgress, withCompletion: completionBlock)
         } else {
            self.animationView?.play(toProgress: toProgress, withCompletion: completionBlock);
         }
      }
      
      
      if(call.method == "playWithFrames") {
         let toFrame = props["toFrame"] as! NSNumber;
         if let fromFrame = props["fromFrame"] as? NSNumber {
            self.animationView?.play(fromFrame: fromFrame, toFrame: toFrame, withCompletion: completionBlock);
         } else {
            self.animationView?.play(toFrame: toFrame, withCompletion: completionBlock);
         }
      }
      
      if(call.method == "stop") {
         self.animationView?.stop();
      }
      
      if(call.method == "pause") {
         self.animationView?.pause();
      }
      
      if(call.method == "setAnimationSpeed") {
         self.animationView?.animationSpeed = props["speed"] as! CGFloat
      }
      
      if(call.method == "setLoopAnimation") {
         self.animationView?.loopAnimation = props["loop"] as! Bool
      }
      
      if(call.method == "setAutoReverseAnimation") {
         self.animationView?.autoReverseAnimation = props["reverse"] as! Bool
      }
      
      if(call.method == "setAnimationProgress") {
         self.animationView?.animationProgress = props["progress"] as! CGFloat
      }
      
      if(call.method == "setProgressWithFrame") {
         let frame = props["frame"] as! NSNumber
         self.animationView?.setProgressWithFrame(frame)
      }
      
      if(call.method == "isAnimationPlaying") {
         let isAnimationPlaying = self.animationView?.isAnimationPlaying
         result(isAnimationPlaying)
      }
      
      if(call.method == "getAnimationDuration") {
         let animationDuration = self.animationView?.animationDuration
         result(animationDuration)
      }
      
      if(call.method == "getAnimationProgress") {
         let animationProgress = self.animationView?.animationProgress
         result(animationProgress)
      }
      
      if(call.method == "getAnimationSpeed") {
         let animationSpeed = self.animationView?.animationSpeed
         result(animationSpeed)
      }
      
      if(call.method == "getLoopAnimation") {
         let loopAnimation = self.animationView?.loopAnimation
         result(loopAnimation)
      }
      
      if(call.method == "getAutoReverseAnimation") {
         let autoReverseAnimation = self.animationView?.autoReverseAnimation
         result(autoReverseAnimation)
      }
      
      
      if(call.method == "setValue") {
         let value = props["value"] as! String;
         let keyPath = props["keyPath"] as! String;
         if let type = props["type"] as? String {
            setValue(type: type, value: value, keyPath: keyPath)
         }
      }
      
   }
   
   func setValue(type: String, value: String, keyPath: String) -> Void {
      switch type {
      case "LOTColorValue":
         let i = UInt32(value.dropFirst(2), radix: 16)
         let color = hexToColor(hex8: i!);
         self.delegates.append(ColorDelegate(color: color))
         self.animationView?.setValueDelegate(self.delegates[self.delegates.count - 1], for: LOTKeypath(string: keyPath + ".Color"))
         break;
      case "LOTOpacityValue":
         if let n = NumberFormatter().number(from: value) {
            let f = CGFloat(truncating: n)
            self.delegates.append(NumberDelegate(number: f))
            self.animationView?.setValueDelegate(self.delegates[self.delegates.count - 1], for: LOTKeypath(string: keyPath + ".Opacity"))
         }
         break;
      default:
         break;
      }
   }
   
}
